### PR TITLE
MLIR: Firsts commits to support LLVM 10 

### PR DESCRIPTION
**Pull request by Robertorosmaninho**

### CODE REVIEW

**Comment by thewilsonator** [comment]
Please remove the unnecessary whitespace changes.

File: driver/codegenerator.cpp [hunk 1]
```diff
@@ -69,14 +74,16 @@ createAndSetDiagnosticsOutputFile(IRState &irs, llvm::LLVMContext &ctx,
...
+    diagnosticsOutputFile = make_unique<llvm::ToolOutputFile>(
         diagnosticsFilename, EC, llvm::sys::fs::F_None);
     if (EC) {
       irs.dmodule->error("Could not create file %s: %s",
                          diagnosticsFilename.c_str(), EC.message().c_str());
       fatal();
     }
 
+    //FIXME: this has been replaced by setDiagnosticHandler
+#if LDC_LLVM_VER < 1000
```
Comment (thewilsonator) [inline driver/codegenerator.cpp]: This looks to be missing a corresponding `#endif`

File: gen/pgo_ASTbased.cpp [hunk 2]
```diff
@@ -888,8 +893,7 @@ void CodeGenPGO::loadRegionCounts(llvm::IndexedInstrProfReader *PGOReader,
     return;
   }
 
-  ProfRecord =
-      llvm::make_unique<llvm::InstrProfRecord>(std::move(RecordExpected.get()));
+  ProfRecord = make_unique<llvm::InstrProfRecord>(std::move(RecordExpected.get()));
```
Comment (thewilsonator) [inline gen/pgo_ASTbased.cpp]: Please leave this as is

File: gen/optimizer.cpp [hunk 3]
```diff
@@ -194,7 +201,9 @@ static void addGarbageCollect2StackPass(const PassManagerBuilder &builder,
 static void addAddressSanitizerPasses(const PassManagerBuilder &Builder,
                                       PassManagerBase &PM) {
   PM.add(createAddressSanitizerFunctionPass());
+
 #if LDC_LLVM_VER >= 900
+
```
Comment (thewilsonator) [inline gen/optimizer.cpp]: You still have a bunch of whitespace.

File: gen/optimizer.cpp [hunk 4]
```diff
@@ -234,12 +243,19 @@ static void addThreadSanitizerPass(const PassManagerBuilder &Builder,
 static void addSanitizerCoveragePass(const PassManagerBuilder &Builder,
                                      legacy::PassManagerBase &PM) {
 #ifdef ENABLE_COVERAGE_SANITIZER
+
 #if LDC_LLVM_VER >= 1000
   PM.add(createModuleSanitizerCoverageLegacyPassPass(
       opts::getSanitizerCoverageOptions()));
 #else
+
```
Comment (thewilsonator) [inline gen/optimizer.cpp]: ditto

File: gen/optimizer.cpp [hunk 5]
```diff
@@ -306,7 +322,9 @@ static void addOptimizationPasses(legacy::PassManagerBase &mpm,
     builder.Inliner = createAlwaysInlinerPass();
 #endif
   }
+
 #if LDC_LLVM_VER < 900
+
```
Comment (thewilsonator) [inline gen/optimizer.cpp]: ditto

File: gen/optimizer.cpp [hunk 6]
```diff
@@ -471,7 +489,9 @@ void outputOptimizationSettings(llvm::raw_ostream &hash_os) {
   hash_os << disableSimplifyDruntimeCalls;
   hash_os << disableSimplifyLibCalls;
   hash_os << disableGCToStack;
+
 #if LDC_LLVM_VER < 900
+
```
Comment (thewilsonator) [inline gen/optimizer.cpp]: ditto

File: gen/passes/GarbageCollect2Stack.cpp [hunk 7]
```diff
@@ -79,13 +79,15 @@ void EmitMemSet(IRBuilder<> &B, Value *Dst, Value *Val, Value *Len,
...
 
+
 #if LDC_LLVM_VER >= 1000
   MaybeAlign Align(1);
 #else
   unsigned Align = 1; 
 #endif
 
   auto CS = B.CreateMemSet(Dst, Val, Len, Align, false /*isVolatile*/);
+
```
Comment (thewilsonator) [inline gen/passes/GarbageCollect2Stack.cpp]: ditto

File: gen/passes/GarbageCollect2Stack.cpp [hunk 8]
```diff
@@ -448,17 +450,18 @@ static void RemoveCall(CallSite CS, const Analysis &A) {
...
     BranchInst::Create(Invoke->getNormalDest(), Invoke);
     Invoke->getUnwindDest()->removePredecessor(CS->getParent());
   }
 
   // Remove the runtime call.
   if (A.CGNode) {
+
 #if LDC_LLVM_VER >= 900
     A.CGNode->removeCallEdgeFor(*cast<CallBase>(CS.getInstruction()));
+
```
Comment (thewilsonator) [inline gen/passes/GarbageCollect2Stack.cpp]: ditto

File: gen/runtime.cpp [hunk 9]
```diff
@@ -343,12 +343,14 @@ llvm::Function *getRuntimeFunction(const Loc &loc, llvm::Module &target,
...
 
   LLFunction *resfn =
+
       llvm::cast<llvm::Function>(target
                                      .getOrInsertFunction(name, fnty)
 #if LDC_LLVM_VER >= 900
                                      .getCallee()
 #endif
       );
+
```
Comment (thewilsonator) [inline gen/runtime.cpp]: ditto

File: driver/codegenerator.cpp (hunk 1 above)
Comment (Robertorosmaninho) [reply driver/codegenerator.cpp]: Indeed

File: gen/pgo_ASTbased.cpp [hunk 10]
```diff
@@ -34,6 +34,11 @@
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/MD5.h"
 
+#if LDC_LLVM_VER >= 1000
+using std::make_unique;
+#else
+using llvm::make_unique;
+#endif
```
Comment (thewilsonator) [inline gen/pgo_ASTbased.cpp]: this looks unused

File: gen/pgo_ASTbased.cpp (hunk 2 above)
Comment (Robertorosmaninho) [reply gen/pgo_ASTbased.cpp]: better now?

File: gen/pgo_ASTbased.cpp (hunk 10 above)
Comment (Robertorosmaninho) [reply gen/pgo_ASTbased.cpp]: Ok, I'll delete this.

File: gen/optimizer.cpp [hunk 11]
```diff
@@ -239,7 +243,12 @@ static void addSanitizerCoveragePass(const PassManagerBuilder &Builder,
       opts::getSanitizerCoverageOptions()));
 #else
   PM.add(
-      createSanitizerCoverageModulePass(opts::getSanitizerCoverageOptions()));
+#if LDC_LLVM_VER >= 1000
+      createModuleSanitizerCoverageLegacyPassPass
+#else
+      createSanitizerCoverageModulePass
+#endif
```
Comment (kinke) [inline gen/optimizer.cpp]: This is useless, as the #else in line 244 is for LLVM < 10 only.

File: driver/codegenerator.cpp [hunk 12]
```diff
@@ -36,6 +36,11 @@ namespace llvm {
 }
 #endif
 
+#if LDC_LLVM_VER >= 1000
+using std::make_unique;
+#else
+using llvm::make_unique;
+#endif
```
Comment (kinke) [inline driver/codegenerator.cpp]: This should also be working already, see https://github.com/ldc-developers/ldc/blob/2ff7d493480aa6c658c6ccd636b23c093906b128/gen/llvm.h#L35-L42. Your work is most likely based on an older LDC head...

File: driver/codegenerator.cpp [hunk 13]
```diff
@@ -69,17 +74,19 @@ createAndSetDiagnosticsOutputFile(IRState &irs, llvm::LLVMContext &ctx,
...
+    diagnosticsOutputFile = make_unique<llvm::ToolOutputFile>(
         diagnosticsFilename, EC, llvm::sys::fs::F_None);
     if (EC) {
       irs.dmodule->error("Could not create file %s: %s",
                          diagnosticsFilename.c_str(), EC.message().c_str());
       fatal();
     }
 
+    //FIXME: this has been replaced by setDiagnosticHandler
+#if LDC_LLVM_VER < 1000
```
Comment (kinke) [inline driver/codegenerator.cpp]: This is also never hit, the #else in line 75 is for LLVM < 9.

**Comment by kinke** [comment]
AFAICT, nothing of this should be required.

**Pull request closed without merging**